sympy/sets/fancysets.py

```diff
@@ -910,14 +910,17 @@ def _boundary(self):
 
     def as_relational(self, x):
         """Rewrite a Range in terms of equalities and logic operators. """
-        from sympy.functions.elementary.integers import floor
         if self.size == 1:
             return Eq(x, self[0])
         else:
+            from sympy.core.mod import Mod
+            from sympy.functions.elementary.integers import floor
+            md = self.start % self.step
```

Comment:
I would use something like the following to compute `md`:
```python
  md =Piecewise(
    (1, Eq(self.start, oo) & Eq(self.stop, -oo) | Eq(self.start, -oo) & Eq(self.stop, oo)),
    (self.reversed.start % self.step, Eq(self.start, -oo)),
    (self.reversed.start % -self.step, Eq(self.start, oo)),
    (self.start % abs(self.step), True))
```
And then the condition to add would be `Eq(Mod(x,abs(self.step)), md)` if both start and stop are not infinite, else True (if both are infinite). I would add an extra `elif self.size == 0` and return `S.false` in that case since that is the empty set.

Here are some test cases:
```python
Range(oo, -oo, -1) True since Eq(x,1) == 0 when x is integer
Range(-oo, oo, 1) True since Eq(x,1) == 0 when x is integer
Range(1, -oo, -2) Eq(Mod(x, 2), 1)
Range(2, -oo, -2) Eq(Mod(x, 2), 0)
Range(6, oo, 2) Eq(Mod(x, 2), 0)
Range(5, oo, 2) Eq(Mod(x, 2), 1)
Range(3, 11, 2) Eq(Mod(x, 2), 1)
Range(4, 10, 2) Eq(Mod(x, 2), 0)
Range(10, 2, -2) Eq(Mod(x, 2), 0)
Range(11, 1, -2) Eq(Mod(x, 2), 1)
Range(0, 0, 1) has size 0 and should return S.false
Range(i, i + 30, 2) Eq(Mod(x, 2), Mod(i, 2))
Range(i + 1, i + 30, 2) Eq(Mod(x, 2), Mod(i + 1, 2))
Range(i, i - 30, -2) Eq(Mod(x, 2), Mod(i, 2))
Range(i + 1, i - 30, -2) Eq(Mod(x, 2), Mod(i + 1, 2))
ValueError for Range(x, x + 30, 2)
```

Reply:
Why is the last one ValueError?

Reply:
> I would use something like the following to compute `md`:

I am editing the code, keeping all the conditions in mind. And I have changed it as :
```
def as_relational(self, x):
    """Rewrite a Range in terms of equalities and logic operators. """
    if self.size == 1:
        return Eq(x, self[0])
    elif self.size == 0:
        return S.false
    else:
        from sympy.functions.elementary.piecewise import Piecewise
        from sympy.core.mod import Mod
        from sympy.functions.elementary.integers import floor
        md = Piecewise(
            (1, Eq(self.start, oo) & Eq(self.stop, -oo) | Eq(self.start, -oo) & Eq(self.stop, oo)),
            (self.reversed.start % self.step, Eq(self.start, -oo)),
            (self.reversed.start % abs(self.step), Eq(self.start, oo)),
            (self.start % abs(self.step), True))
        if Eq(self.start, oo) & Eq(self.stop, -oo) | Eq(self.start, -oo) & Eq(self.stop, oo):
            return True
        elif Eq(self.start, -oo):
            return And(
                Eq(x, floor(x)),
                x >= -oo,
                x <= self.sup if self.sup in self else x < self.sup,
                Eq(Mod(x, abs(self.step)), md))
        elif Eq(self.start, oo):
            return And(
                Eq(x, floor(x)),
                x >= self.inf if self.inf in self else x > self.inf,
                x <= oo,
                Eq(Mod(x, abs(self.step)), md))
        else:
            return And(
                Eq(x, floor(x)),
                x >= self.inf if self.inf in self else x > self.inf,
                x <= self.sup if self.sup in self else x < self.sup,
                Eq(Mod(x, abs(self.step)), md))
```


> Why is the last one ValueError?


This was due to the condition in the property <code>size(self)</code>
the present code is :
```

    @property
    def size(self):
        if not self:
            return S.Zero
        dif = self.stop - self.start
        if self.has(Symbol):
            if dif.has(Symbol) or self.step.has(Symbol) or (
                    not self.start.is_integer and not self.stop.is_integer):
                raise ValueError('invalid method for symbolic range')
        if dif.is_infinite:
            return S.Infinity
        return Integer(abs(dif//self.step))
```
I am  changing the code to this :
```
    @property
    def size(self):
        if not self:
            return S.Zero
        if self.has(Symbol):
            if self.step.has(Symbol) or (
                    not self.start.is_integer and not self.stop.is_integer):
                raise ValueError('invalid method for symbolic range')
        dif = self.stop - self.start
        if dif.is_infinite:
            return S.Infinity
        return Integer(abs(dif//self.step))
```
This codes works if you set the symbols (starting and ending for range) as integer and it won't give any error now.

>```
> Range(i, i + 30, 2) Eq(Mod(x, 2), Mod(i, 2))
> Range(i + 1, i + 30, 2) Eq(Mod(x, 2), Mod(i + 1, 2))
> Range(i, i - 30, -2) Eq(Mod(x, 2), Mod(i, 2))
> Range(i + 1, i - 30, -2) Eq(Mod(x, 2), Mod(i + 1, 2))
> ValueError for Range(x, x + 30, 2)
> ```

I executed all of these tests they seems to work fine now. If there is still any issue I will work on it to resolve.

Reply:
> Why is the last one ValueError?

it fails at the `size` check since it is not known that x is an integer.

Reply:
> I am editing the code, keeping all the conditions in mind.

it is a shame to lose the simplicity of your original. This exercise teaches that we really only need to make sure the `md` picks a valid point:

```
if self.start.is_infinite:
    if self.stop.is_infinite:
        a = 0
    else:
        a = self.reversed.start
elif self.start != self.stop:
    a = self.start
else:
    return S.false # for null range
...
include condition `Eq(Mod(x, self.step), a % self.step)` and let the signs remain.

Your modifications to `size` look good.

Reply:
But we might as well keep the extra negative out of the step. Something like this?
```python
        if self.size == 1:
            return Eq(x, self[0])
        elif self.size == 0:
            return S.false
        else:
            from sympy.core.mod import Mod
            from sympy.functions.elementary.integers import floor
            cond = None
            if self.start.is_infinite:
                if self.stop.is_infinite:
                    cond = S.true
                else:
                    a = self.reversed.start
            elif self.start == self.stop:
                cond = S.false  # null range
            else:
                a = self.start
            step = abs(self.step)
            cond = Eq(Mod(x, step), a % step) if cond is None else cond
            return And(
                cond, Eq(x, floor(x)),
                x >= self.inf if self.inf in self else x > self.inf,
                x <= self.sup if self.sup in self else x < self.sup)
```

Reply:
> it is a shame to lose the simplicity of your original. This exercise teaches that we really only need to make sure the `md` picks a valid point:

I will keep that in mind to not to lose the originality from next time on.
Sir as for now I have added  `Contains`  to the code you have modified. I tried to edit the code but I could not come up with a better way than yours. 
It works properly as it is :
```
        if self.size == 1:
            return Eq(x, self[0])
        elif self.size == 0:
            return S.false
        else:
            from sympy.core.mod import Mod
            from sympy.sets.contains import Contains
            cond = None
            if self.start.is_infinite:
                if self.stop.is_infinite:
                    cond = S.true
                else:
                    a = self.reversed.start
            elif self.start == self.stop:
                cond = S.false  # null range
            else:
                a = self.start
            step = abs(self.step)
            cond = Eq(Mod(x, step), a % step) if cond is None else cond
            return And(Contains(x,S.Integers),
                        cond,
                        x >= self.inf if self.inf in self else x > self.inf,
                        x <= self.sup if self.sup in self else x < self.sup)
````
I executed all tests that you have mentioned along with some extra tests they all are working properly without giving any error.

Reply:
Sounds good! So just push your branch with these changes using `git push Sagar231` from the branch you are working in.